ClassCastException with code action and method part

Given this Qute template:

```
{@java.util.List<String> strings}
{#let name=strings.get(XXX)}
	
{/let}
```

XXX is higlighted as an error. if you hove it (to exectute code action), you can see in the trace:

```
[Error - 09:27:13] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.ClassCastException: class com.redhat.qute.parser.expression.MethodPart cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.MethodPart and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.ClassCastException: class com.redhat.qute.parser.expression.MethodPart cannot be cast to class com.redhat.qute.parser.expression.ObjectPart (com.redhat.qute.parser.expression.MethodPart and com.redhat.qute.parser.expression.ObjectPart are in unnamed module of loader 'app')
	at com.redhat.qute.services.codeactions.QuteCodeActionForUndefinedObject.doCodeActions(QuteCodeActionForUndefinedObject.java:77)
```

--- a/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/utils/QutePositionUtility.java
+++ b/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/utils/QutePositionUtility.java
@@ -145,7 +145,7 @@ public class QutePositionUtility {
 					expression = section.getExpressionParameter();
 				}
 				if (expression != null) {
-					Node expressionNode = expression.findNodeExpressionAt(offset);
+					Node expressionNode = findBestNode(expression, offset);
 					if (expressionNode != null) {
 						return expressionNode;
 					}

--- a/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/codeaction/QuteCodeActionForSimilarTextSuggestionsTest.java
+++ b/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/codeaction/QuteCodeActionForSimilarTextSuggestionsTest.java
@@ -102,6 +102,32 @@ public class QuteCodeActionForSimilarTextSuggestionsTest {
 	}
 
 	@Test
+	public void smilarTextSuggestionQuickFixForUndefinedObjectInMethodPart() throws Exception {
+		String template = "{@java.util.List<org.acme.Item> items}\r\n" + //
+				"{@java.lang.Integer index}\r\n" + //
+				"{#let name=items.get(inde)}\r\n" + //
+				"	\r\n" + //
+				"{/let}";
+
+		Diagnostic d = d(2, 21, 2, 25, //
+				QuteErrorCode.UndefinedObject, //
+				"`inde` cannot be resolved to an object.", //
+				DiagnosticSeverity.Warning);
+
+		testDiagnosticsFor(template, d);
+		testCodeActionsFor(template, d, //
+				ca(d, te(2, 21, 2, 25, "index")), //
+				ca(d, te(0, 0, 0, 0, "{@java.lang.String inde}\r\n")), //
+				ca(d, te(2, 25, 2, 25, "??")), //
+				ca(d, c("Ignore `UndefinedObject` problem.", //
+						QuteClientCommandConstants.COMMAND_CONFIGURATION_UPDATE, //
+						"qute.validation.undefinedObject.severity", //
+						"test.qute", //
+						ConfigurationItemEditType.update, "ignore", //
+						d)));
+	}
+
+	@Test
 	public void similarTextSuggestionQuickFixForUndefinedNamespace() throws Exception {
 		String template = "{inje:model}";
 		Diagnostic d = d(0, 1, 0, 5, //
@@ -222,18 +248,18 @@ public class QuteCodeActionForSimilarTextSuggestionsTest {
 
 	@Test
 	public void similarTextSuggestionQuickFixForUndefinedMethodFromExtendedType() throws Exception {
-	String template = "{@org.acme.Item item}\r\n" + //
-			"{item.conver()}";
-	
-	Diagnostic d = d(1, 6, 1, 12, //
-			QuteErrorCode.UnknownMethod, //
-			"`conver` cannot be resolved or is not a method of `org.acme.Item` Java type.", //
-			DiagnosticSeverity.Error);
-	d.setData(new JavaBaseTypeOfPartData("org.acme.Item"));
-	
-	testDiagnosticsFor(template, d);
-	testCodeActionsFor(template, d, //
-			ca(d, te(1, 6, 1, 12, "convert")));
+		String template = "{@org.acme.Item item}\r\n" + //
+				"{item.conver()}";
+
+		Diagnostic d = d(1, 6, 1, 12, //
+				QuteErrorCode.UnknownMethod, //
+				"`conver` cannot be resolved or is not a method of `org.acme.Item` Java type.", //
+				DiagnosticSeverity.Error);
+		d.setData(new JavaBaseTypeOfPartData("org.acme.Item"));
+
+		testDiagnosticsFor(template, d);
+		testCodeActionsFor(template, d, //
+				ca(d, te(1, 6, 1, 12, "convert")));
 	}
 
 	@Test

--- a/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/hover/QuteHoverInExpressionWithLetSectionTest.java
+++ b/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/hover/QuteHoverInExpressionWithLetSectionTest.java
@@ -55,4 +55,28 @@ public class QuteHoverInExpressionWithLetSectionTest {
 				"```", //
 				r(0, 10, 0, 18));
 	}
+
+	@Test
+	public void hoverMethodPart() throws Exception {
+		String template = "{@java.lang.String string}\r\n" + //
+				"{#let name=string.cha|rAt(0) /}";
+		assertHover(template, "```java" + //
+				System.lineSeparator() + //
+				"char java.lang.String.charAt(int index)" + //
+				System.lineSeparator() + //
+				"```", //
+				r(1, 18, 1, 24));
+	}
+	
+	@Test
+	public void hoverParameterOfMethodPart() throws Exception {
+		String template = "{@java.lang.String string}\r\n" + //
+				"{#let name=string.charAt(0|) /}";
+		assertHover(template, "```java" + //
+				System.lineSeparator() + //
+				"Integer" + //
+				System.lineSeparator() + //
+				"```", //
+				r(1, 25, 1, 26));
+	}
 }

